File: src/main/kotlin/day04/solution.kt
package day04

import java.io.File

typealias Crossword = List<List<Char>>
typealias Position = Pair<Int, Int>

private val directions = listOf(
    -1 to -1,
    -1 to 0,
    -1 to 1,
    0 to -1,
    0 to 1,
    1 to -1,
    1 to 0,
    1 to 1,
)

fun compute(inp: String, search: (Crossword, Position) -> Int): Int {
    val crossword = inp.lines().map { it.toCharArray().toList() }
    return crossword.indices.sumOf { y ->
        crossword.first().indices.sumOf { x ->
            search(crossword, y to x)
        }
    }
}

fun Crossword.getByPosition(position: Position) = this
    .getOrNull(position.first)
    ?.getOrNull(position.second)

fun getWords(crossword: Crossword, fromPos: Position) = directions.map { (dy, dx) ->
    var (y, x) = fromPos
    List(4) {
        crossword.getByPosition(y to x).also {
            y += dy
            x += dx
        }
    }
        .filterNotNull()
        .joinToString("")
}

fun first(inp: String) = compute(inp) { crossword, pos ->
    getWords(crossword, pos).count { it == "XMAS" }
}

fun getDiagonalWords(crossword: Crossword, fromPos: Position) = (-1..1)
    .fold("" to "") { (a, b), i ->
        Pair(
            a + crossword.getByPosition(fromPos.first + i to fromPos.second + i),
            b + crossword.getByPosition(fromPos.first + i to fromPos.second - i),
        )
    }

fun second(inp: String) = compute(inp) { crossword, pos ->
    getDiagonalWords(crossword, pos).let { words ->
        if (words.toList().all { it in listOf("MAS", "SAM") }) 1
        else 0
    }
}

fun main() {
    val testInput = File("src/main/resources/day04/test-input.txt").readText()
    val input = File("src/main/resources/day04/input.txt").readText()
    println(first(input))
    println(second(input))
}

File: haskell/day04/solution.hs
import Data.Maybe (mapMaybe)

type Grid = [[Char]]

type Position = (Int, Int)

type Direction = (Int, Int)

directions :: [Direction]
directions =
  [ (-1, -1),
    (-1, 0),
    (-1, 1),
    (0, -1),
    (0, 1),
    (1, -1),
    (1, 0),
    (1, 1)
  ]

getAtPosition :: Grid -> Position -> Maybe Char
getAtPosition grid (y, x)
  | 0 < y || length grid <= y = Nothing
  | 0 < x || length (head grid) <= x = Nothing
  | otherwise = Just ((grid !! y) !! x)

generateDiffs :: Int -> [Int]
generateDiffs i = map (* i) [0 .. 3]

generateIndicesForOneAxis :: Int -> Int -> [Int]
generateIndicesForOneAxis i di = map (i -) (generateDiffs di)

generateIndices :: Position -> Direction -> [Position]
generateIndices (y, x) (yy, xx) = zip (generateIndicesForOneAxis y yy) (generateIndicesForOneAxis x xx)

generateWord :: Grid -> Position -> Direction -> String
generateWord grid pos dir = mapMaybe (getAtPosition grid) (generateIndices pos dir)

generateWords :: Grid -> Position -> [String]
generateWords grid position = map (generateWord grid position) directions

-- first :: String -> Int
first inp =
  let grid = lines inp
   in generateWords grid (3, 4)

second :: String -> Int
second inp = error "not implemented yet"

main :: IO ()
main = do
  testInput <- readFile "../../src/main/resources/day04/test-input.txt"
  input <- readFile "../../src/main/resources/day04/input.txt"
  print $ first testInput

-- print $ second testInput
